how_to_eurec4a/cloudmasks.md

```diff
@@ -4,136 +4,261 @@ jupytext:
 \begin{equation}
 height = height\_above\_sea\_level[0] - range
 \end{equation}
 
 ```{code-cell} ipython3
-def add_height_coordinate(ds):
+def add_height_coordinate_wales(ds):
     ds.coords["height"] = ds.height_above_sea_level[0].values - ds.range
-    return ds
+    return ds.swap_dims({"range": "height"})
 ```
 
 ```{code-cell} ipython3
-ds_bsrg = add_height_coordinate(cat.HALO.WALES.bsrg["HALO-0205"].to_dask().load())
-ds_bsri = add_height_coordinate(cat.HALO.WALES.bsri["HALO-0205"].to_dask().load())
+ds_bsri = add_height_coordinate_wales(cat.HALO.WALES.bsri["HALO-0205"].to_dask())
+da_bsri = ds_bsri.backscatter_ratio.sel(height=slice(4000, 0), time=s)
 ```
 
-#### Cloud top height
+From WALES we also include the cloud top height information
 
 ```{code-cell} ipython3
-ds_cloud = cat.HALO.WALES.cloudparameter["HALO-0205"].to_dask().load()
+da_cth = cat.HALO.WALES.cloudparameter["HALO-0205"].to_dask().cloud_top.sel(time=s)
 ```
 
-## Plots
+#### SpecMACS imager radiance
+From specMACS we include the radiances at 1.6 micron in the short-wave infrared (SWIR). Note: this sepcMACS radiance dataset is only available for the following application on February 5, not for the whole campaign.
 
 ```{code-cell} ipython3
-from mpl_toolkits.axes_grid1 import make_axes_locatable
-from matplotlib.colors import LogNorm
+url = ("https://observations.ipsl.fr/thredds/dodsC/EUREC4A/PRODUCTS/SPECMACS-CLOUDMASK/"
+       + "EUREC4A_HALO_specMACS_cloud_mask_20200205T100000-20200205T182359_v1.1.nc")
+ds_swir = xr.open_dataset(url, engine="netcdf4")
+da_swir = ds_swir.sel(time=s).isel(radiation_wavelength=0).swir_radiance
 ```
 
-We choose an example time interval
+#### Preprocess cloud mask data
+
+We copy the data dictionary and apply the time selection.
+
++++
+
+For the 2D horizontal imagers we select a region in the center, derive a representative `cloud_mask` flag value and use that in the following intercomparison plot.
+* VELOX: we select only the cetral 10 x 10 pixels
+* SpecMACS: we select the central 0.6 degrees, i.e. angle = 0 ∓ 0.3
 
 ```{code-cell} ipython3
-s = slice(datetime.datetime(2020, 2, 5, 11, 22, 0), datetime.datetime(2020, 2, 5, 11, 27, 0))
+def most_frequent(da):
+    flag, n = np.unique(da.values, return_counts=True)
+    return xr.DataArray(int(flag[np.argmax(n)]), dims=())
+
+def select_specmacs_cloudmask(ds):
+    specmacs = ds.sel(angle=slice(.3, -.3))
+    return ds.assign({"cloud_mask": specmacs.cloud_mask.groupby("time").map(most_frequent),
+                      "CF_min": cfmin(specmacs),
+                      "CF_max": cfmax(specmacs)})
+
+def select_velox_cloudmask(ds):
+    xmid = ds.x.size // 2
+    ymid = ds.y.size // 2
+    velox = ds.isel(x=slice(xmid - 5, xmid + 5), y=slice(ymid - 5, ymid + 5))
+    return ds.assign({"cloud_mask": velox.cloud_mask.groupby("time").map(most_frequent),
+                      "CF_min": cfmin(velox),
+                      "CF_max": cfmax(velox)})
```

Comment:
This one should provide quite a bit of speed-up.

Take-home-message: `groupby` is really slow...
```suggestion
def most_frequent_flag(var, dims):
    flags = xr.DataArray(var.flag_values, dims=("__internal__flags__"))
    flag_indices = (var == flags).sum(dims).argmax("__internal__flags__")
    return xr.DataArray(var.flag_values[flag_indices.data],
                        dims=flag_indices.dims,
                        attrs=var.attrs)

def select_specmacs_cloudmask(ds):
    specmacs = ds.sel(angle=slice(.3, -.3))
    return ds.assign({"cloud_mask": most_frequent_flag(specmacs.cloud_mask, "angle"),
                      "CF_min": cfmin(specmacs),
                      "CF_max": cfmax(specmacs)})

def select_velox_cloudmask(ds):
    xmid = ds.x.size // 2
    ymid = ds.y.size // 2
    velox = ds.isel(x=slice(xmid - 5, xmid + 5), y=slice(ymid - 5, ymid + 5))
    return ds.assign({"cloud_mask": most_frequent_flag(velox.cloud_mask, ("x", "y")),
                      "CF_min": cfmin(velox),
                      "CF_max": cfmax(velox)})
```